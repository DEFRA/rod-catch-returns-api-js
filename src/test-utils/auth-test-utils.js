import fetch from 'node-fetch'
import { getSystemUserByOid } from '../services/system-users.service.js'
import jwksClient from 'jwks-rsa'
import jwt from 'jsonwebtoken'

jest.mock('jsonwebtoken')
jest.mock('jwks-rsa')
jest.mock('../services/system-users.service.js')

export const getMockAuthAndUser = (userOverrides) => {
  fetch.mockResolvedValueOnce({
    json: () =>
      Promise.resolve({
        jwks_uri: 'https://example.com/jwks'
      }),
    ok: true
  })

  jwksClient.mockReturnValueOnce({
    getSigningKey: jest.fn().mockResolvedValue({
      publicKey: 'mock-public-key'
    })
  })

  jwt.decode.mockReturnValueOnce({ header: { kid: 'abc' } })
  jwt.verify.mockReturnValueOnce({ oid: 'abc12345' })

  getSystemUserByOid.mockResolvedValueOnce({
    isDisabled: false,
    roles: [],
    ...userOverrides
  })
}
