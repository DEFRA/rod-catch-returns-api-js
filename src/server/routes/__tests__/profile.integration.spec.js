import { getMockAuthAndUser } from '../../../test-utils/auth-test-utils.js'
import initialiseServer from '../../server.js'

describe('profile.integration', () => {
  /** @type {import('@hapi/hapi').Server} */
  let server = null

  beforeAll(async () => {
    server = await initialiseServer()
  })

  afterAll(async () => {
    await server.stop()
  })

  describe('GET /api/profile', () => {
    it('should return a list of all the urls available in the api', async () => {
      const result = await server.inject({
        method: 'GET',
        url: '/api/profile'
      })

      expect(result.statusCode).toBe(200)
      expect(JSON.parse(result.payload)).toMatchSnapshot()
    })

    it("should return a 403 if the user's account is disabled", async () => {
      getMockAuthAndUser({
        isDisabled: true,
        roles: [{ name: 'RCR CRM Integration User' }]
      })

      const result = await server.inject({
        method: 'GET',
        url: '/api/profile',
        headers: {
          token: 'abc123'
        }
      })

      expect(result.statusCode).toBe(403)
      expect(JSON.parse(result.payload)).toStrictEqual({
        error: 'ACCOUNT_DISABLED'
      })
    })

    it('should return a 403 if the user has an incorrect role', async () => {
      getMockAuthAndUser({
        isDisabled: false,
        roles: [{ name: 'Some Other Role' }]
      })

      const result = await server.inject({
        method: 'GET',
        url: '/api/profile',
        headers: {
          token: 'abc123'
        }
      })

      expect(result.statusCode).toBe(403)
      expect(JSON.parse(result.payload)).toStrictEqual({
        error: 'ACCOUNT_ROLE_REQUIRED'
      })
    })

    it.each(['RCR CRM Integration User', 'System Administrator'])(
      'should return a 200 if the user has a %s role',
      async (role) => {
        getMockAuthAndUser({
          isDisabled: false,
          roles: [{ name: role }]
        })

        const result = await server.inject({
          method: 'GET',
          url: '/api/profile',
          headers: {
            token: 'abc123'
          }
        })

        expect(result.statusCode).toBe(200)
      }
    )
  })
})
