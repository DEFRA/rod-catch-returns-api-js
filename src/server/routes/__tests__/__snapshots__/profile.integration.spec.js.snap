// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile.integration GET /api/profile should return a list of all the urls available in the api 1`] = `
{
  "urls": [
    {
      "method": "GET",
      "path": "/documentation",
    },
    {
      "method": "GET",
      "path": "/service-status",
    },
    {
      "method": "GET",
      "path": "/swagger.json",
    },
    {
      "method": "GET",
      "path": "/{param*}",
    },
    {
      "method": "GET",
      "path": "/api/catchments",
    },
    {
      "method": "GET",
      "path": "/api/grilseProbabilities",
    },
    {
      "method": "GET",
      "path": "/api/grilseWeightGates",
    },
    {
      "method": "GET",
      "path": "/api/methods",
    },
    {
      "method": "GET",
      "path": "/api/profile",
    },
    {
      "method": "GET",
      "path": "/api/regions",
    },
    {
      "method": "GET",
      "path": "/api/rivers",
    },
    {
      "method": "GET",
      "path": "/api/species",
    },
    {
      "method": "GET",
      "path": "/swaggerui/extend.js",
    },
    {
      "method": "GET",
      "path": "/swaggerui/favicon-16x16.png",
    },
    {
      "method": "GET",
      "path": "/swaggerui/favicon-32x32.png",
    },
    {
      "method": "GET",
      "path": "/swaggerui/index.html",
    },
    {
      "method": "GET",
      "path": "/swaggerui/oauth2-redirect.html",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui-bundle.js",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui-bundle.js.map",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui-standalone-preset.js",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui-standalone-preset.js.map",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui.css",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui.css.map",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui.js",
    },
    {
      "method": "GET",
      "path": "/swaggerui/swagger-ui.js.map",
    },
    {
      "method": "GET",
      "path": "/api/activities/{activityId}",
    },
    {
      "method": "GET",
      "path": "/api/catches/{catchId}",
    },
    {
      "method": "GET",
      "path": "/api/licence/{licence}",
    },
    {
      "method": "GET",
      "path": "/api/methods/{methodId}",
    },
    {
      "method": "GET",
      "path": "/api/smallCatches/{smallCatchId}",
    },
    {
      "method": "GET",
      "path": "/api/submissions/{submissionId}",
    },
    {
      "method": "GET",
      "path": "/api/activities/{activityId}/submission",
    },
    {
      "method": "GET",
      "path": "/api/activities/{activityId}/catches",
    },
    {
      "method": "GET",
      "path": "/api/activities/{activityId}/river",
    },
    {
      "method": "GET",
      "path": "/api/activities/{activityId}/smallCatches",
    },
    {
      "method": "GET",
      "path": "/api/catches/{catchId}/species",
    },
    {
      "method": "GET",
      "path": "/api/catches/{catchId}/activity",
    },
    {
      "method": "GET",
      "path": "/api/catches/{catchId}/method",
    },
    {
      "method": "GET",
      "path": "/api/licence/full/{licence}",
    },
    {
      "method": "GET",
      "path": "/api/smallCatches/{smallCatchId}/activity",
    },
    {
      "method": "GET",
      "path": "/api/submissions/search/findByContactId",
    },
    {
      "method": "GET",
      "path": "/api/submissions/search/getByContactIdAndSeason",
    },
    {
      "method": "GET",
      "path": "/api/submissions/{submissionId}/activities",
    },
    {
      "method": "GET",
      "path": "/api/reporting/reference/grilse-probabilities/{season}",
    },
    {
      "method": "POST",
      "path": "/api/activities",
    },
    {
      "method": "POST",
      "path": "/api/catches",
    },
    {
      "method": "POST",
      "path": "/api/smallCatches",
    },
    {
      "method": "POST",
      "path": "/api/submissions",
    },
    {
      "method": "POST",
      "path": "/api/reporting/reference/grilse-probabilities/{season}/{gate}",
    },
    {
      "method": "DELETE",
      "path": "/api/activities/{activityId}",
    },
    {
      "method": "DELETE",
      "path": "/api/catches/{catchId}",
    },
    {
      "method": "DELETE",
      "path": "/api/grilseProbabilities/{grilseProbabilityId}",
    },
    {
      "method": "DELETE",
      "path": "/api/smallCatches/{smallCatchId}",
    },
    {
      "method": "DELETE",
      "path": "/api/submissions/{submissionId}",
    },
    {
      "method": "PATCH",
      "path": "/api/activities/{activityId}",
    },
    {
      "method": "PATCH",
      "path": "/api/catches/{catchId}",
    },
    {
      "method": "PATCH",
      "path": "/api/smallCatches/{smallCatchId}",
    },
    {
      "method": "PATCH",
      "path": "/api/submissions/{submissionId}",
    },
    {
      "method": "PUT",
      "path": "/api/catches/{catchId}/activity",
    },
    {
      "method": "PUT",
      "path": "/api/smallCatches/{smallCatchId}/activity",
    },
  ],
}
`;
